zadanie 1
a) Podziel w losowy sposób bazę danych na zbiór treningowy (70%) i testowy
(30%).

In [266]:
import pandas as pd
from sklearn.model_selection import train_test_split

df = pd.read_csv("iris.csv")

X = df.iloc[:, :-1].values
y = df.iloc[:, 4].values

#podzial na zbior testowy (30%) i treningowy (70%), ziarno losowosci = 274929
X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.7, random_state=274929)

b) Uruchom każdy z klasyfikatorów wykorzystując paczki i dokonaj ewaluacji ma
zbiorze testowym wyświetlając procentową dokładność i macierz błędu.

1. k-NN, k=3

In [267]:
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report, confusion_matrix

scaler = StandardScaler()
scaler.fit(X_train)

X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)

classifier = KNeighborsClassifier(n_neighbors=3)
classifier.fit(X_train, y_train)

y_pred = classifier.predict(X_test)

print(confusion_matrix(y_test, y_pred))

numberOfAll = X_test.shape[0]
numberOfWrong = (y_test != y_pred).sum()
numberOfCorrect = numberOfAll - numberOfWrong

print("Procent dobrze sklasyfikowanych irysów %d" % (numberOfCorrect/numberOfAll * 100))

[[16  1  0]
 [ 0 14  1]
 [ 0  1 12]]
Procent dobrze sklasyfikowanych irysów 93


Otrzymany wynik to 93%

2. k-NN, k=5

In [268]:
classifier = KNeighborsClassifier(n_neighbors=5)
classifier.fit(X_train, y_train)

y_pred = classifier.predict(X_test)

print(confusion_matrix(y_test, y_pred))

numberOfAll = X_test.shape[0]
numberOfWrong = (y_test != y_pred).sum()
numberOfCorrect = numberOfAll - numberOfWrong

print("Procent dobrze sklasyfikowanych irysów %d" % (numberOfCorrect/numberOfAll * 100))

[[17  0  0]
 [ 0 14  1]
 [ 0  1 12]]
Procent dobrze sklasyfikowanych irysów 95


Otrzymany wynik to 95%

3. k-NN, k=11

In [269]:
classifier = KNeighborsClassifier(n_neighbors=11)
classifier.fit(X_train, y_train)

y_pred = classifier.predict(X_test)

print(confusion_matrix(y_test, y_pred))

numberOfAll = X_test.shape[0]
numberOfWrong = (y_test != y_pred).sum()
numberOfCorrect = numberOfAll - numberOfWrong

print("Procent dobrze sklasyfikowanych irysów %d" % (numberOfCorrect/numberOfAll * 100))

[[17  0  0]
 [ 0 12  3]
 [ 0  2 11]]
Procent dobrze sklasyfikowanych irysów 88


Otrzymany wynik to 88%

zadanie 2
Dokonaj klasyfikacji zbioru danych z irysami za pomocą Naive Bayesa

In [270]:
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB

df = pd.read_csv("iris.csv")

X = df.iloc[:, :-1].values
y = df.iloc[:, 4].values

X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.7, random_state=274709)

gnb = GaussianNB()
y_pred = gnb.fit(X_train, y_train).predict(X_test)

numberOfAll = X_test.shape[0]
numberOfWrong = (y_test != y_pred).sum()
numberOfCorrect = numberOfAll - numberOfWrong

print("Procent %d"% (numberOfCorrect/numberOfAll * 100))

Procent 91


Wynik: 91% irysów zostało dobrze sklasyfikowanych.

zad 3
a) Załadować paczkę sklearn , bazę danych z irysami I podzielić irysy na część
testową i treningową używając komendy: train_test_split ( 70% / 30% ).

In [285]:
df = pd.read_csv("iris.csv")

X = df.iloc[:, :-1].values
y = df.iloc[:, 4].values

X_train, X_test, y_train, y_test = train_test_split(X, y, train_size=0.7, random_state=274779)

b) Sieć neuronowa nie akceptuje napisów, jedynie liczby. Może się nauczyć
rozpoznawać irysy, pod warunkiem że ich nazwy są numeryczne np. 0, 1 lub 2.
Trzeba więc zamienić nazwy tekstowe na liczby przed treningiem.

In [286]:
target = df[['class']].replace(['setosa','versicolor','virginica'],[0,1,2])
# print(target.sample(n=5))

Jakie to liczby? Jakim napisom odpowiadają?
* setosa - 0
* versicolor - 1
* virginica - 2

c) Możesz przeskalować dane, ale ten punkt nie jest obowiązkowy

In [287]:
df_norm = df[['sepallength', 'sepalwidth', 'petallength', 'petalwidth']].apply(lambda x: (x - x.min()) / (x.max() - x.min()))

d) Skonstruuj I wytrenuj model sieci neuronowej z czteroneuronową warstwą
wejściową (bo są cztery pomiary irysów), jedną ukrytą warstwą z dwoma
neuronami i warstwą wyjściową z jednym neuronem decydującym o gatunku
irysa.

In [288]:
import pandas as pd
from sklearn.neural_network import MLPClassifier # neural network
from sklearn import metrics

df = pd.concat([df_norm, target], axis=1)
# df.sample(n=5)

train, test = train_test_split(df, train_size=0.7, random_state=274779)

trainX = train[['sepallength','sepalwidth','petallength','petalwidth']]# taking the training data features

trainY = train['class'] # output of our training data

testX= test[['sepallength','sepalwidth','petallength','petalwidth']] # taking test data features

testY = test['class'] # output value of test data

clf = MLPClassifier(solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(2,), random_state=1)

clf.fit(trainX, trainY)

print(trainY)

25     0
128    2
20     0
54     1
140    2
      ..
119    2
97     1
39     0
24     0
67     1
Name: class, Length: 105, dtype: int64


e) Wyjaśnij jak liczby zwracane przez sieć (są to liczby rzeczywiste,
zmiennoprzecinkowe) zamieniane są na odpowiedzi (gatunki irysa).

wynik jest zaokrąglany.


In [289]:
prediction = clf.predict(testX)
# print(testX)
# print(testY)
print("-")
print(prediction)

-
[0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0]


f) Dokonaj ewaluacji sieci na zbiorze testowym. Wyświetl jej dokładność.

In [290]:
print('The accuracy of the Multi-layer Perceptron is:',metrics.accuracy_score(prediction,testY)) #(2,)

The accuracy of the Multi-layer Perceptron is: 0.24444444444444444


Wyszło 28.8%

g) Sprawdź czy model sieci z trzema neuronami działa lepiej.

In [291]:
clf = MLPClassifier(solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(3,), max_iter=1000, random_state=1)
clf.fit(trainX, trainY)



prediction = clf.predict(testX)
# print(prediction)

print('The accuracy of the Multi-layer Perceptron is:',metrics.accuracy_score(prediction,testY))

The accuracy of the Multi-layer Perceptron is: 0.9333333333333333


Działa lepiej, wyszło 88.9%

h) Sprawdź czy model z dwiema warstwami neuronowymi, po 3 neurony każda,
działa lepiej.

In [292]:
clf = MLPClassifier(solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(3, 3), random_state=1)

clf.fit(trainX, trainY)

prediction = clf.predict(testX)
print(prediction)

print('The accuracy of the Multi-layer Perceptron is:',metrics.accuracy_score(prediction,testY))

[2 1 2 0 1 2 2 1 0 1 1 0 1 2 2 2 0 1 1 2 1 2 2 2 2 0 2 2 2 2 2 1 0 1 0 1 0
 2 0 1 0 2 2 1 0]
The accuracy of the Multi-layer Perceptron is: 0.9777777777777777


Działa jeszcze lepiej, wyszło: 93.4%

i) Innym sposobem odgadywania irysów jest przyporządkowanie każdej wartości
klasy jednemu neuronowi wyjściowemu (np. górny to setoa, środkowy
versicolor, dolny virginica).

In [302]:
options = ["setosa", "versicolor", "virginica"]

def name_to_list(name):
    return [int(c == name) for c in options]

# zamiast zamieniac nazwa_kwiatka -> liczba
# zamieniam nazwa_kwiatka -> [liczba, liczba, liczba]
def class_names_to_lists(classes_names):
    return [name_to_list(c) for c in classes_names]

df = pd.read_csv("iris.csv")

train, test = train_test_split(df, train_size=0.7, random_state=274779)

trainX = train[['sepallength','sepalwidth','petallength','petalwidth']]# taking the training data features
trainY = class_names_to_lists(train['class'])

testX= test[['sepallength','sepalwidth','petallength','petalwidth']] # taking test data features
testY = class_names_to_lists(test['class']) # output value of test data

clf = MLPClassifier(solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(2,), max_iter=1000, random_state=1)
clf.fit(trainX, trainY)



prediction = clf.predict(testX)
#print(prediction)

print(trainY)

print('The accuracy of the Multi-layer Perceptron is:',metrics.accuracy_score(prediction,testY))

[[0, 0, 1], [0, 0, 1], [0, 1, 0], [1, 0, 0], [0, 0, 1], [1, 0, 0], [0, 1, 0], [0, 1, 0], [1, 0, 0], [1, 0, 0], [1, 0, 0], [0, 0, 1], [0, 1, 0], [0, 1, 0], [0, 0, 1], [1, 0, 0], [0, 1, 0], [0, 0, 1], [0, 0, 1], [0, 0, 1], [1, 0, 0], [0, 0, 1], [0, 1, 0], [0, 0, 1], [1, 0, 0], [1, 0, 0], [0, 0, 1], [1, 0, 0], [0, 0, 1], [0, 1, 0], [0, 1, 0], [1, 0, 0], [0, 1, 0], [1, 0, 0], [0, 1, 0], [1, 0, 0], [0, 0, 1], [0, 0, 1], [1, 0, 0], [0, 1, 0], [0, 1, 0], [1, 0, 0], [0, 0, 1], [1, 0, 0], [1, 0, 0], [0, 1, 0], [0, 0, 1], [0, 0, 1], [1, 0, 0], [0, 0, 1], [1, 0, 0], [0, 1, 0], [0, 1, 0], [0, 1, 0], [0, 1, 0], [1, 0, 0], [0, 1, 0], [0, 0, 1], [1, 0, 0], [0, 1, 0], [0, 1, 0], [1, 0, 0], [0, 1, 0], [0, 1, 0], [0, 1, 0], [0, 0, 1], [1, 0, 0], [0, 0, 1], [0, 1, 0], [0, 0, 1], [1, 0, 0], [1, 0, 0], [1, 0, 0], [1, 0, 0], [0, 1, 0], [0, 1, 0], [0, 1, 0], [1, 0, 0], [1, 0, 0], [0, 1, 0], [0, 1, 0], [0, 1, 0], [0, 0, 1], [0, 0, 1], [0, 0, 1], [0, 0, 1], [0, 0, 1], [0, 0, 1], [1, 0, 0], [1, 0, 0], [0, 0, 1]

hidden_layer_sizes=(2,):
* poprzednio wyszło 28.8%
* teraz: 66.7%
czyli wynik się poprawił

hidden_layer_sizes=(3,):

In [298]:
clf = MLPClassifier(solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(3,), max_iter=1000, random_state=1)
clf.fit(trainX, trainY)



prediction = clf.predict(testX)
# print(prediction)

print('The accuracy of the Multi-layer Perceptron is:',metrics.accuracy_score(prediction,testY))

The accuracy of the Multi-layer Perceptron is: 0.6666666666666666


* Poprzednio = 88.9%
* Teraz = 57.8%
Pogorszyło się

hidden_layer_sizes=(3,):

In [295]:
clf = MLPClassifier(solver='lbfgs', alpha=1e-5, hidden_layer_sizes=(3,3), max_iter=1000, random_state=1)
clf.fit(trainX, trainY)



prediction = clf.predict(testX)
# print(prediction)

print('The accuracy of the Multi-layer Perceptron is:',metrics.accuracy_score(prediction,testY))

The accuracy of the Multi-layer Perceptron is: 0.9555555555555556


* Poprzednio = 93.4%
* Teraz = 95.6%
Tym razem się polepszyło.
Czyli raz trudno powiedzieć który sposób jest lepszy.